### ISSUE TITLE
Error: graph model model.predict throws `avoid the dynamic ops` error

### ISSUE THREAD

**Issue by HarshalRohit** [tensorflow/tfjs#4579]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Pop-os 20.10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Any android device (checked on LG G8x, Oneplus 3)
- TensorFlow.js installed from (npm or script link): npm
- TensorFlow.js version (use command below): ^2.7.0
- Browser version: NA
- Tensorflow.js Converter Version: 2.3.0

**Describe the current behavior**
I used tensorflow-js converter to obtain a graph model from keras saved-model.
The conversion is successful and also throw any error when loaded using `tf.loadGraphModel`
but when trying to call `model.predict` it throws following error
```shell
Error: This execution contains the node 'StatefulPartitionedCall/model/rnn_1/while/exit/_65', 
which has the dynamic op 'Exit'. Please use model.executeAsync() instead. 
Alternatively, to avoid the dynamic ops, specify the inputs [Identity]
```

When I try `model.executeAsync()`, following error occurs
```shell
Error: Cannot compute the outputs [Identity] from the provided inputs [the_input]. 
Consider providing the following inputs: []. Alternatively, to avoid the dynamic ops, 
use model.execute() and specify the inputs [Identity]
```

For `model.execute()` similar kind of error is thrown.
I also tried to specify the `outputs` arguments for `execute` and `executeAsync` but nothing seems to work.

My model consists of following layers:
+ InputLayer
+ Conv1D
+ BatchNormalisation
+ GRU
+ TimeDistributed
+ Softmax - Activation

The tensorflow-js converter readme page mentions that inference is faster for `graphModel` which I intend to leverage.

**Describe the expected behavior**
`model.predict` should return a Tensor/TensorArray.

**Comment by rthadur** [comment]
@HarshalRohit can you please provide codepen example to reproduce the same and also can you try with latest version

**Comment by HarshalRohit** [comment]
> @HarshalRohit can you please provide codepen example to reproduce the same and also can you try with latest version

Hi, I tried with updated tensorflowjs as well as tensorflowjs_converter libraries, but same errors. 
My use-case is in React-native, and I don't think codepen supports that.
Following is the code,
```js
import * as tf from '@tensorflow/tfjs';
import {bundleResourceIO} from '@tensorflow/tfjs-react-native';

const modelDirName = 'tmp';
const modelJson = require(`../assets/${modelDirName}/model.json`);
const modelWeights = require(`../assets/${modelDirName}/group1-shard1of1.bin`);

const loadModel = async () => {
  const model = await tf.loadGraphModel(
    bundleResourceIO(modelJson, modelWeights),
  );

  // WARM-UP model
  const res = tf.tidy(() =>
    model.predict(tf.zeros([1, 128, 161])),
  ) as tf.Tensor;                    // <<---------- Error is thrown here

  return model;
};

export {loadModel};
```
Same code for a layers-model (obtained from same .h5 file) works fine. The layers used in the model are mentioned in previous post.

**Comment by mattsoulanille** [comment]
@HarshalRohit thanks for the report. I'm trying to reproduce the error you're seeing, but it looks like it only appears with your model. I ran one of our test models in an example app (https://github.com/mattsoulanille/tfjs-rn-webgl-test/tree/test-model), and it threw no errors. If you can share the model you're working with, that would help me debug this. If not, that's fine too. I'll create one with the layers you specified above.

**Comment by HarshalRohit** [comment]
I will be happy to share the model. But is it possible to share it privately??

**Comment by HarshalRohit** [comment]
@mattsoulanille Any updates??

**Comment by mattsoulanille** [comment]
@HarshalRohit Sorry for the wait. Would you be willing to invite me to a private GitHub repo with the model script in it so I can put it through tfjs-converter and try running it on a device? Feel free to omit the training weights. Thanks!

**Comment by HarshalRohit** [comment]
@mattsoulanille You should have received an invite. 
I have added the required files. 
Please feel free to point out any missing files. Thank you!!

**Comment by mattsoulanille** [comment]
Hi @HarshalRohit. Thanks for posting the model. I was able to reproduce the error you're seeing, and I've pushed my reproduction to the repo you shared. I've also found that if you use a Layers model instead of a Graph model, it works fine (see the layers branch on the repo), so it looks like this could be an issue with tfjs-react-native. I'll take a look.

**Comment by HarshalRohit** [comment]
Hi @mattsoulanille. Thanks for the update. 
Yup, it works for Layers model.
I want to use Graph Model as its supposed to be faster at inference than Layers model.

**Comment by HarshalRohit** [comment]
Hi @mattsoulanille, Any updates??

**Comment by mattsoulanille** [comment]
I've reproduced the bug in the browser, so this seems to be an issue with converter or the graph model executor rather than tfjs-react-native as I originally thought. 

In the browser, the layers model works fine, but when running the graph model, It looks like the last layer (Identity) is not connected to the rest of it. When running with a single input, it throws the same error as was first reported...
```
Error: Cannot compute the outputs [Identity] from the provided inputs [the_input]. 
Consider providing the following inputs: []. Alternatively, to avoid the dynamic ops, 
use model.execute() and specify the inputs [Identity]
```
...and when I provide `Identity` as an input as requested, `executeAsync()` just returns whatever the `Identity` input is, short-circuiting the model. 

This seems like it could be related to https://github.com/tensorflow/tfjs/issues/2263, but that's been fixed.

I'm going to ask someone more familiar with the graph model executor to take a look. Thanks for your patience!

**Comment by mattsoulanille** [comment]
Hi @HarshalRohit. Thanks to @pyu10055, we've found a workaround. If you convert the model with [`--control_flow_v2=true`](https://www.tensorflow.org/api_docs/python/tf/compat/v1/enable_control_flow_v2), it should work. I've pushed a working example to the private repo, and the exact converter command is in the package.json file. To anyone else reading this, the command is essentially:
```bash
tensorflowjs_converter --input_format keras --output_format tfjs_graph_model --control_flow_v2=true model.h5 graphModel/
```

We'll continue investigating the cause of this bug, but hopefully that should unblock you for now.

**Comment by HarshalRohit** [comment]
Thank you for looking into it.
I will try the workaround and update.

**Comment by HarshalRohit** [comment]
The workaround works. 
The prediction time for graph model is worse then the layers model but that's a separate issue :p
Thanks!!!

**Comment by playground** [comment]
@mattsoulanille I tried the work around by adding --control_flow_v2=True 

```tensorflowjs_converter \
    --input_format=tf_saved_model \
    --output_format=tfjs_graph_model \
    --saved_model_tags=serve \
    --control_flow_v2=True \
    --signature_name=serving_default \
    /savedmodel
    /TFConvertModel
```

I get this error

```2021-07-10 16:24:51.666225: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2021-07-10 16:24:51.705248: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x108c297d0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2021-07-10 16:24:51.705291: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version

TypeError: Cannot read property 'children' of undefined
    at /object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:3678:33
    at Array.forEach (<anonymous>)
    at /object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:3675:29
    at Array.forEach (<anonymous>)
    at OperationMapper.mapFunction (/object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:3673:18)
    at /object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:3507:56
    at Array.reduce (<anonymous>)
    at OperationMapper.transformGraph (/object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:3506:48)
    at GraphModel.loadSync (/object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:7411:68)
    at GraphModel.<anonymous> (/object-detection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:7385:52)```

**Comment by sickopickle** [comment]
I tried to add control_flow_v2=true, but I now get this error: This execution contains the node 'StatefulPartitionedCall/model/bidirectional_2/backward_rnn_2/while', which has the dynamic op 'While'. Please use model.executeAsync() instead. Alternatively, to avoid the dynamic ops, specify the inputs [StatefulPartitionedCall/model/bidirectional_2/backward_rnn_2/TensorArrayV2Stack/TensorListStack]

**Comment by sickopickle** [comment]
@pyu10055 Can you help me? I tried to use a LayersModel at first, but I am using a SlicingOpLambda which is not supported. I tried to use GraphModel, but I get these errors. I thought it might be because of absl: untraced functions warning, so I tried doing this suggestion: https://github.com/tensorflow/tensorflow/issues/47554#issuecomment-1204201567, which took away the warning. Originally, the error was `This execution contains the node 'StatefulPartitionedCall/model/bidirectional_2/forward_lstm_2/PartitionedCall/while/exit/_370', which has the dynamic op 'Exit'. Please use model.executeAsync() instead. Alternatively, to avoid the dynamic ops, specify the inputs [StatefulPartitionedCall/model/bidirectional_2/forward_lstm_2/PartitionedCall/TensorArrayV2Stack/TensorListStack]`, but after doing this, the error changed to the error described in this post: `This execution contains the node 'StatefulPartitionedCall/model/rnn_1/while/exit/_65', 
which has the dynamic op 'Exit'. Please use model.executeAsync() instead. 
Alternatively, to avoid the dynamic ops, specify the inputs [Identity]`. Even after using control_flow_v2, I still get dynamic op errors. :(
